File: src/main/java/nl/hanze/week2/ATM/ATM.java
package nl.hanze.week2.ATM;

import nl.hanze.week2.ATM.proxy.AbstractAtm;

// the context
public class ATM implements AbstractAtm {
    private ATMState state;
    private int amount = 100;

    // define an initial state
    public ATM() {
        setState(new IdleState());
    }

    void setState(ATMState state) {
        this.state = state;
    }

    public ATMState getState() {
        return this.state;
    }

    @Override
    public int getCashInMachine() {
        return amount;
    }

    void insertCard() {
        state.insertCard(this);
    }

    void ejectCard() {
        state.ejectCard(this);
    }

    void insertPin() {
        state.insertPin(this);
    }

    void requestAmount() {
        state.requestAmount(this);
    }
}

File: src/main/java/nl/hanze/week2/assignment4/ATM.kt
//// the context
//class assignment45 {
//    internal var state: ATMState? = null
//    private val amount = 100
//
//    // define an initial state
//    init {
//        state = IdleState()
//    }
//
//    internal fun insertCard() {
//        state!!.insertCard(this)
//    }
//
//    internal fun ejectCard() {
//        state!!.ejectCard(this)
//    }
//
//    internal fun insertPin() {
//        state!!.insertPin(this)
//    }
//
//    internal fun requestAmount() {
//        state!!.requestAmount(this)
//    }
//}
